### PR TITLE
Api authentication

### PR DESCRIPTION
**Pull request by nbrabant**
Hello,
I've just create the authentication process for the admin panel with simple auth with InMemoryUserProvider
Don't hesitate to give me feedback and review to improve the job
Regards

### CODE REVIEW

File: api/src/Controller/AuthController.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+{
+
+    public function login_check(Request $request)
+    {
+        // set the in_memory provider
+        $userProvider = new InMemoryUserProvider(
+            array(
+                'admin' => array(
+                    // password is "foo"
+                    'password' => '5FZ2Z8QIkA7UTZ4BYkoC+GsReLf569mSKDsfods6LYQ8t+a8EW9oaircfMpmaLbPBh4FOBiiFyLfuZmTSUwzZg==',
```
Comment (alanpoulain) [inline api/src/Controller/AuthController.php]: If you use the `PlaintextPasswordEncoder` (see below), you don't need the hash.

File: api/src/Controller/AuthController.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                ),
+            )
+        );
+
+        // encoder array of password encoders (see below)
+        $encoderFactory = new EncoderFactory([
+            User::class => new MessageDigestPasswordEncoder('sha512', true, 5000)
+        ]);
+
+        // define the data object provider
```
Comment (alanpoulain) [inline api/src/Controller/AuthController.php]: You can remove this comment.

File: api/src/Controller/AuthController.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            $request->request->get('username'),
+            $request->request->get('password'),
+            'secured_area'
+        );
+
+        return $this->json([
+            'token' => $daoProvider->authenticate($unauthenticatedToken)
+        ]);
+    }
+
```
Comment (alanpoulain) [inline api/src/Controller/AuthController.php]: This line can be removed.

File: api/src/Controller/AuthController.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+class AuthController extends AbstractController
+{
+
+    public function login_check(Request $request)
+    {
+        // set the in_memory provider
+        $userProvider = new InMemoryUserProvider(
+            array(
```
[the remark this answers is not in the document]
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: Ok, aslo for admin configuration?

File: api/src/Controller/AuthController.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Security\Core\Encoder\MessageDigestPasswordEncoder;
+use Symfony\Component\Security\Core\User\User;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+class AuthController extends AbstractController
+{
+
+    public function login_check(Request $request)
+    {
+        // set the in_memory provider
```
[the remark this answers is not in the document]
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok

File: api/src/Controller/AuthController.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Security\Core\User\UserChecker;
+use Symfony\Component\Security\Core\User\InMemoryUserProvider;
+use Symfony\Component\Security\Core\Encoder\EncoderFactory;
+use Symfony\Component\Security\Core\Encoder\MessageDigestPasswordEncoder;
+use Symfony\Component\Security\Core\User\User;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+class AuthController extends AbstractController
+{
+
```
[the remark this answers is not in the document]
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok

File: api/src/Controller/AuthController.php (hunk 3 above)
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok

File: api/src/Controller/AuthController.php (hunk 2 above)
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok

File: api/src/Controller/AuthController.php (hunk 1 above)
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok, cool for tests ^^

File: api/src/Controller/AuthController.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                    // password is "foo"
+                    'password' => '5FZ2Z8QIkA7UTZ4BYkoC+GsReLf569mSKDsfods6LYQ8t+a8EW9oaircfMpmaLbPBh4FOBiiFyLfuZmTSUwzZg==',
+                    'roles'    => array('ROLE_ADMIN'),
+                ),
+            )
+        );
+
+        // encoder array of password encoders (see below)
+        $encoderFactory = new EncoderFactory([
+            User::class => new MessageDigestPasswordEncoder('sha512', true, 5000)
```
[the remark this answers is not in the document]
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok

File: api/src/Controller/AuthController.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Security\Core\User\InMemoryUserProvider;
+use Symfony\Component\Security\Core\Encoder\EncoderFactory;
+use Symfony\Component\Security\Core\Encoder\MessageDigestPasswordEncoder;
+use Symfony\Component\Security\Core\User\User;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+class AuthController extends AbstractController
+{
+
+    public function login_check(Request $request)
```
[the remark this answers is not in the document]
Comment (nbrabant) [reply api/src/Controller/AuthController.php]: ok

**Comment by nbrabant** [comment]
Thanks for the review

File: api/src/Controller/AuthController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanpoulain) [reply api/src/Controller/AuthController.php]: Sure, everywhere :wink:

**Comment by dunglas** [comment]
You could only just the config: https://symfony.com/doc/current/security/multiple_user_providers.html
Alternatively, we may use the new maker bundle commands: https://symfony.com/blog/new-in-makerbundle-1-8-instant-user-login-form-commands

**Comment by vincentchalamon** [comment]
You can use the [Symfony JSON login setup](https://symfony.com/doc/current/security/json_login_setup.html). Here is a complete implementation, including logout & cookie http-only + secure:
```php
namespace App\Controller;

use ApiPlatform\Core\JsonLd\Serializer\ItemNormalizer;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
use Symfony\Component\Serializer\SerializerInterface;

final class Security
{
    /**
     * @Route(name="api_login_check", path="/login_check", methods={"POST"})
     */
    public function __invoke(SerializerInterface $serializer, TokenStorageInterface $tokenStorage): JsonResponse
    {
        return new JsonResponse($serializer->serialize($tokenStorage->getToken()->getUser(), ItemNormalizer::FORMAT), JsonResponse::HTTP_OK, [
            'Content-Type' => 'application/ld+json',
        ], true);
    }
}
```
```yaml
# config/routes/annotations.yaml
controllers:
    resource: ../../src/Controller/
    type: annotation
```
```yaml
# config/packages/security.yaml
security:
    encoders:
        App\Entity\User: bcrypt

    # https://symfony.com/doc/current/security.html#where-do-users-come-from-user-providers
    providers:
        api_user_provider:
            entity:
                class: App\Entity\User

    role_hierarchy:
        ROLE_ADMIN: [ROLE_USER]

    firewalls:
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false

        main:
            anonymous: true
            json_login:
                check_path: /login
                require_previous_session: true
            logout:
                path: logout
```
```yaml
# config/routes/security.yaml
logout:
    path: /logout
```
```yaml
# config/packages/framework.yaml
framework:
    session:
        cookie_httponly: true
        cookie_secure: true
        name: APIPLATFORMSESSID
```

**Comment by nbrabant** [comment]
Hello and thanks for your review
Unfortunately, i don't have much time for now to do this improvements. It's in my Trello dev's Todolist board ;)

**Comment by teohhanhui** [comment]
Replaced by #111. I didn't realize there was already an open PR. Sorry.

If you have time, do help review. Thanks @nbrabant!

**Pull request closed without merging**